### DISCUSSION TITLE
[serializer] MAX_SIZE of StreamOutput?

### DISCUSSION THREAD

**Discussion by schlangz** [activej/activej#174] [Q&A]
Hi, the serializer is great stuff I enjoy working with it. Question on the maximum size of StreamOutput:
`private static final int MAX_SIZE = 1 << 28; // 256MB`
Is the 256MB defined on purpose or just because it shouldn't be unlimited? Thank you

**Comment by arsentiikaptsan** [comment]
Could I ask what I think is a related question: is it possible to get a ChannelSupplier in RPC request handler, so I could process large data as it arrives (+ benefit from backpressure)? Or maybe RPC was not meant for such kind of tasks and I should use HTTP or your fs lib instead? Thanks!

**Comment by eduard-vasinskyi** [comment]
Hello, @schlangz
[StreamOutput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/stream/StreamOutput.java) uses (1-4 byte) VarInt encoding to serialize the size of each stream item. This allows saving space when serializing a lot of small items. Maximum value of a 4-byte VarInt is exactly 256MB. The serializer is intended primarily for processing a large number of small messages of the same type, such as logs, RPC requests/responses, etc.

**Comment by eduard-vasinskyi** [comment]
Hello, @arsentiikaptsan
RPC was designed for quick request/response communication (millions of requests per second). It is not suited for working with large data streams.
As for processing large data, you can use the [HTTP module](https://github.com/activej/activej/tree/master/core-http) (it is a separate module, not part of [ActiveJ FS](https://github.com/activej/activej/tree/master/cloud-fs)). By using HTTP request/response body [ChannelSupplier](https://github.com/activej/activej/blob/master/core-csp/src/main/java/io/activej/csp/ChannelSupplier.java) you can process request data as a stream, one piece at a time.
A more advanced solution is to use [MessagingWithBinaryStreaming](https://github.com/activej/activej/blob/master/core-csp/src/main/java/io/activej/csp/net/MessagingWithBinaryStreaming.java) class that works on top of a raw TCP socket and allows working with socket data using CSP. You will have to design some protocol to exchange data between peers. You may take a look at [ActiveFsServer](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/tcp/ActiveFsServer.java) and [RemoteActiveFs](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/tcp/RemoteActiveFs.java) as a reference. Although, we recommend sticking to the HTTP solution. ActiveJ provides one of the fastest Java HTTP client/server, which should be enough for most common applications. Turn to design your own protocol only if you are sure you need to squeeze every last drop of performance.

**Comment by schlangz** [comment]
> Hello, @schlangz
> [StreamOutput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/stream/StreamOutput.java) uses (1-4 byte) VarInt encoding to serialize the size of each stream item. This allows saving space when serializing a lot of small items. Maximum value of a 4-byte VarInt is exactly 256MB. The serializer is intended primarily for processing a large number of small messages of the same type, such as logs, RPC requests/responses, etc.

Many of my objects are bigger, is there any chance of getting a StreamOutput that can handle larger object sizes? I'm really in favor of ActiveJ serialization because of the simplicity and amazing speed it provides.

**Comment by eduard-vasinskyi** [comment]
@schlangz 
You could use [ActiveJ Serializer](https://serializer.activej.io/) without `StreamInput/Output`. Just create a [BinarySerializer](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinarySerializer.java) and use it to serialize data to a byte array. You can look at [serializer examples](https://github.com/activej/activej/tree/master/examples/core/serializer/src/main/java) for reference.
